widget/list.go

```diff
@@ -0,0 +1,405 @@
+	list := &List{BaseWidget: BaseWidget{}, Length: length, CreateItem: createItem, UpdateItem: updateItem, selectedIndex: -1}
+	list.ExtendBaseWidget(list)
+	return list
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (l *List) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if l.itemMin.IsZero() && l.CreateItem != nil {
+		l.itemMin = newListItem(l.CreateItem(), nil).MinSize()
+	}
+	layout := fyne.NewContainerWithLayout(newListLayout(l))
+	layout.Resize(layout.MinSize())
+
+	scroller := NewVScrollContainer(layout)
+
+	scroller.onOffsetChanged = func() {
+		if l.offsetY == scroller.Offset.Y {
+			return
+		}
+		l.offsetY = scroller.Offset.Y
+		l.BaseWidget.Refresh()
+	}
+
+	objects := []fyne.CanvasObject{scroller}
+	return &listRenderer{BaseRenderer: widget.NewBaseRenderer(objects), list: l, scroller: scroller, layout: layout}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (l *List) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+type listRenderer struct {
+	widget.BaseRenderer
+	list             *List
+	scroller         *ScrollContainer
+	layout           *fyne.Container
+	itemPool         *listPool
+	children         []fyne.CanvasObject
+	visibleItemCount int
+	firstItemIndex   int
+	lastItemIndex    int
+	size             fyne.Size
+}
+
+func (l *listRenderer) Layout(size fyne.Size) {
+	if size != l.size {
+		if size.Width != l.size.Width {
+			for _, child := range l.children {
+				child.Resize(fyne.NewSize(size.Width, l.list.itemMin.Height))
+			}
+		}
+		l.scroller.Resize(size)
+		l.size = size
+	}
+	if l.list.Length() == 0 {
+		return
+	}
+	if l.itemPool == nil {
+		l.itemPool = &listPool{}
+	}
+	offsetChange := l.list.previousOffsetY - l.list.offsetY
+
+	if offsetChange < 0 {
+		// Scrolling Down
+		l.scrollDown(offsetChange)
+		return
+
+	} else if offsetChange > 0 {
+		// Scrolling Up
+		l.scrollUp(offsetChange)
+		return
+	}
+
+	// Relayout What Is Visible - no scroll change - initial layout or possibly from a resize
+	for _, child := range l.children {
+		l.itemPool.Release(child)
+	}
+	l.children = nil
+	l.visibleItemCount = int(math.Ceil(float64(l.scroller.size.Height) / float64(l.list.itemMin.Height)))
+	if len(l.children) >= l.visibleItemCount || len(l.children) >= l.list.Length() {
+		return
+	}
+	for i := len(l.children) + l.firstItemIndex; len(l.children) <= l.visibleItemCount && i < l.list.Length(); i++ {
+		l.appendItem(i)
+	}
+	l.layout.Objects = l.children
+	l.layout.Layout.Layout(l.layout.Objects, l.list.itemMin)
+	l.lastItemIndex = l.firstItemIndex + len(l.children) - 1
+
+	poolCapacity := 5
+	if l.list.Length()-len(l.children) < 5 {
+		poolCapacity = l.list.Length() - len(l.children)
+	}
+	for i := l.itemPool.Count(); i < poolCapacity; i++ {
+		// Make sure to keep extra items in the pool
```

Comment:
Each time I read an inline comment I think "should this be another method with a clear name".
Should this, with a few lines above, be in an "ensurePoolCapacity" method or something like that?

Reply:
Agreed - Resolved